File: crates/tenzro-training/src/quantization.rs
//! Blockwise symmetric quantization for outer-gradient payloads.
//!
//! Streaming DiLoCo (arXiv 2501.18512) shows outer gradients tolerate 4-bit
//! quantization with no convergence loss — they are momentum-smoothed deltas,
//! not raw activations. This module is the Rust side of the wire format; the
//! Python reference trainer (`tenzro_trainer.quantization`) implements the
//! byte-identical encoder/decoder.
//!
//! # Wire format
//!
//! The tensor is split into blocks of `block_size` values (the final block
//! may be shorter). Each block serializes as:
//!
//! - `Int8`: 4-byte little-endian `f32` scale, then one `i8` code per value.
//!   `scale = max_abs / 127`, `code = round(value / scale)` clamped to
//!   `[-127, 127]`.
//! - `Int4`: 4-byte little-endian `f32` scale, then `ceil(len / 2)` bytes of
//!   codes packed two per byte (low nibble first, odd tail zero-padded).
//!   `scale = max_abs / 7`, codes clamped to `[-7, 7]`, stored as 4-bit
//!   two's complement.
//!
//! All-zero blocks store `scale = 0.0` and all-zero codes; dequantization of
//! a zero scale yields zeros. Dequantized value = `code as f32 * scale`.
//!
//! `GradientQuantization::None` is raw little-endian `f32` — the pre-existing
//! payload format, unchanged.

use tenzro_types::training::GradientQuantization;

use crate::error::{Result, TrainingError};

/// Encode a slice of `f32` values into the quantized wire format.
pub fn quantize(values: &[f32], quantization: GradientQuantization) -> Vec<u8> {
    match quantization {
        GradientQuantization::None => {
            let mut out = Vec::with_capacity(values.len() * 4);
            for v in values {
                out.extend_from_slice(&v.to_le_bytes());
            }
            out
        }
        GradientQuantization::Int8 { block_size } => {
            let block = block_size.max(1) as usize;
            let mut out = Vec::with_capacity(values.len() + values.len().div_ceil(block) * 4);
            for chunk in values.chunks(block) {
                let max_abs = chunk.iter().fold(0.0f32, |acc, v| acc.max(v.abs()));
                let scale = max_abs / 127.0;
                out.extend_from_slice(&scale.to_le_bytes());
                for v in chunk {
                    let code = if scale == 0.0 {
                        0i8
                    } else {
                        (v / scale).round().clamp(-127.0, 127.0) as i8
                    };
                    out.push(code as u8);
                }
            }
            out
        }
        GradientQuantization::Int4 { block_size } => {
            let block = block_size.max(1) as usize;
            let mut out =
                Vec::with_capacity(values.len() / 2 + values.len().div_ceil(block) * 5);
            for chunk in values.chunks(block) {
                let max_abs = chunk.iter().fold(0.0f32, |acc, v| acc.max(v.abs()));
                let scale = max_abs / 7.0;
                out.extend_from_slice(&scale.to_le_bytes());
                let mut iter = chunk.iter();
                loop {
                    let Some(lo) = iter.next() else { break };
                    let hi = iter.next();
                    let lo_code = encode_nibble(*lo, scale);
                    let hi_code = hi.map(|v| encode_nibble(*v, scale)).unwrap_or(0);
                    out.push((lo_code & 0x0F) | (hi_code << 4));
                }
            }
            out
        }
    }
}

/// Decode a quantized wire payload back into `f32` values. `expected_len`
/// is the element count declared by the task's architecture — a payload
/// whose structure disagrees with it is rejected as malformed.
pub fn dequantize(
    bytes: &[u8],
    quantization: GradientQuantization,
    expected_len: usize,
) -> Result<Vec<f32>> {
    match quantization {
        GradientQuantization::None => {
            if bytes.len() != expected_len * 4 {
                return Err(TrainingError::QuantizedPayloadMalformed(format!(
                    "raw f32 payload is {} bytes, expected {} for {} values",
                    bytes.len(),
                    expected_len * 4,
                    expected_len
                )));
            }
            Ok(bytes
                .chunks_exact(4)
                .map(|c| f32::from_le_bytes([c[0], c[1], c[2], c[3]]))
                .collect())
        }
        GradientQuantization::Int8 { block_size } => {
            let block = block_size.max(1) as usize;
            let num_blocks = expected_len.div_ceil(block);
            let expected_bytes = expected_len + num_blocks * 4;
            if bytes.len() != expected_bytes {
                return Err(TrainingError::QuantizedPayloadMalformed(format!(
                    "int8/{block_size} payload is {} bytes, expected {expected_bytes} for \
                     {expected_len} values",
                    bytes.len()
                )));
            }
            let mut out = Vec::with_capacity(expected_len);
            let mut cursor = 0usize;
            let mut remaining = expected_len;
            while remaining > 0 {
                let block_len = remaining.min(block);
                let scale = f32::from_le_bytes([
                    bytes[cursor],
                    bytes[cursor + 1],
                    bytes[cursor + 2],
                    bytes[cursor + 3],
                ]);
                cursor += 4;
                for _ in 0..block_len {
                    out.push(bytes[cursor] as i8 as f32 * scale);
                    cursor += 1;
                }
                remaining -= block_len;
            }
            Ok(out)
        }
        GradientQuantization::Int4 { block_size } => {
            let block = block_size.max(1) as usize;
            let num_blocks = expected_len.div_ceil(block);
            // Each block: 4-byte scale + ceil(block_len / 2) code bytes.
            let mut expected_bytes = num_blocks * 4;
            let mut rem = expected_len;
            while rem > 0 {
                let block_len = rem.min(block);
                expected_bytes += block_len.div_ceil(2);
                rem -= block_len;
            }
            if bytes.len() != expected_bytes {
                return Err(TrainingError::QuantizedPayloadMalformed(format!(
                    "int4/{block_size} payload is {} bytes, expected {expected_bytes} for \
                     {expected_len} values",
                    bytes.len()
                )));
            }
            let mut out = Vec::with_capacity(expected_len);
            let mut cursor = 0usize;
            let mut remaining = expected_len;
            while remaining > 0 {
                let block_len = remaining.min(block);
                let scale = f32::from_le_bytes([
                    bytes[cursor],
                    bytes[cursor + 1],
                    bytes[cursor + 2],
                    bytes[cursor + 3],
                ]);
                cursor += 4;
                let code_bytes = block_len.div_ceil(2);
                for i in 0..block_len {
                    let byte = bytes[cursor + i / 2];
                    let nibble = if i % 2 == 0 { byte & 0x0F } else { byte >> 4 };
                    out.push(decode_nibble(nibble) as f32 * scale);
                }
                cursor += code_bytes;
                remaining -= block_len;
            }
            Ok(out)
        }
    }
}

/// Size in bytes of the quantized encoding of `len` values.
pub fn encoded_len(len: usize, quantization: GradientQuantization) -> usize {
    match quantization {
        GradientQuantization::None => len * 4,
        GradientQuantization::Int8 { block_size } => {
            let block = block_size.max(1) as usize;
            len + len.div_ceil(block) * 4
        }
        GradientQuantization::Int4 { block_size } => {
            let block = block_size.max(1) as usize;
            let mut total = len.div_ceil(block) * 4;
            let mut rem = len;
            while rem > 0 {
                let block_len = rem.min(block);
                total += block_len.div_ceil(2);
                rem -= block_len;
            }
            total
        }
    }
}

fn encode_nibble(value: f32, scale: f32) -> u8 {
    let code = if scale == 0.0 {
        0i8
    } else {
        (value / scale).round().clamp(-7.0, 7.0) as i8
    };
    // 4-bit two's complement.
    (code as u8) & 0x0F
}

fn decode_nibble(nibble: u8) -> i8 {
    // Sign-extend a 4-bit two's-complement value.
    if nibble & 0x08 != 0 {
        (nibble | 0xF0) as i8
    } else {
        nibble as i8
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn sample(n: usize) -> Vec<f32> {
        // Deterministic pseudo-gradient values spanning signs and magnitudes.
        (0..n)
            .map(|i| {
                let x = i as f32;
                ((x * 0.7311).sin() * 0.05) + ((x * 0.1123).cos() * 0.002)
            })
            .collect()
    }

    #[test]
    fn none_round_trips_exactly() {
        let values = sample(1000);
        let bytes = quantize(&values, GradientQuantization::None);
        assert_eq!(bytes.len(), 4000);
        let back = dequantize(&bytes, GradientQuantization::None, 1000).unwrap();
        assert_eq!(back, values);
    }

    #[test]
    fn int8_round_trip_error_bounded() {
        let values = sample(1000);
        let q = GradientQuantization::Int8 { block_size: 256 };
        let bytes = quantize(&values, q);
        assert_eq!(bytes.len(), encoded_len(1000, q));
        let back = dequantize(&bytes, q, 1000).unwrap();
        assert_eq!(back.len(), values.len());
        for (chunk_v, chunk_b) in values.chunks(256).zip(back.chunks(256)) {
            let max_abs = chunk_v.iter().fold(0.0f32, |a, v| a.max(v.abs()));
            let bound = max_abs / 127.0 / 2.0 + f32::EPSILON;
            for (v, b) in chunk_v.iter().zip(chunk_b) {
                assert!(
                    (v - b).abs() <= bound * 1.01,
                    "int8 error {} exceeds half-step {}",
                    (v - b).abs(),
                    bound
                );
            }
        }
    }

    #[test]
    fn int4_round_trip_error_bounded() {
        let values = sample(999); // odd length exercises nibble padding
        let q = GradientQuantization::Int4 { block_size: 128 };
        let bytes = quantize(&values, q);
        assert_eq!(bytes.len(), encoded_len(999, q));
        let back = dequantize(&bytes, q, 999).unwrap();
        assert_eq!(back.len(), values.len());
        for (chunk_v, chunk_b) in values.chunks(128).zip(back.chunks(128)) {
            let max_abs = chunk_v.iter().fold(0.0f32, |a, v| a.max(v.abs()));
            let bound = max_abs / 7.0 / 2.0 + f32::EPSILON;
            for (v, b) in chunk_v.iter().zip(chunk_b) {
                assert!(
                    (v - b).abs() <= bound * 1.01,
                    "int4 error {} exceeds half-step {}",
                    (v - b).abs(),
                    bound
                );
            }
        }
    }

    #[test]
    fn int4_compresses_roughly_8x() {
        let q = GradientQuantization::Int4 { block_size: 256 };
        let raw = encoded_len(100_000, GradientQuantization::None);
        let packed = encoded_len(100_000, q);
        assert!(
            (raw as f64 / packed as f64) > 7.5,
            "ratio {}",
            raw as f64 / packed as f64
        );
    }

    #[test]
    fn zero_block_round_trips_to_zeros() {
        let values = vec![0.0f32; 300];
        for q in [
            GradientQuantization::Int8 { block_size: 128 },
            GradientQuantization::Int4 { block_size: 128 },
        ] {
            let bytes = quantize(&values, q);
            let back = dequantize(&bytes, q, 300).unwrap();
            assert!(back.iter().all(|v| *v == 0.0), "{q}");
        }
    }

    #[test]
    fn odd_tail_block_shorter_than_block_size() {
        let values = sample(130); // 128 + 2 tail
        let q = GradientQuantization::Int8 { block_size: 128 };
        let bytes = quantize(&values, q);
        assert_eq!(bytes.len(), encoded_len(130, q));
        let back = dequantize(&bytes, q, 130).unwrap();
        assert_eq!(back.len(), 130);
    }

    #[test]
    fn wrong_length_rejected() {
        let values = sample(100);
        for q in [
            GradientQuantization::None,
            GradientQuantization::Int8 { block_size: 64 },
            GradientQuantization::Int4 { block_size: 64 },
        ] {
            let bytes = quantize(&values, q);
            let err = dequantize(&bytes, q, 101).unwrap_err();
            assert!(
                matches!(err, TrainingError::QuantizedPayloadMalformed(_)),
                "{q}: {err:?}"
            );
            let err = dequantize(&bytes[..bytes.len() - 1], q, 100).unwrap_err();
            assert!(
                matches!(err, TrainingError::QuantizedPayloadMalformed(_)),
                "{q}: {err:?}"
            );
        }
    }

    #[test]
    fn saturating_values_clamp() {
        // One outlier per block saturates the code range without panicking.
        let mut values = vec![0.001f32; 64];
        values[0] = 10.0;
        for q in [
            GradientQuantization::Int8 { block_size: 64 },
            GradientQuantization::Int4 { block_size: 64 },
        ] {
            let bytes = quantize(&values, q);
            let back = dequantize(&bytes, q, 64).unwrap();
            assert!((back[0] - 10.0).abs() < 0.01, "{q}: outlier {}", back[0]);
        }
    }
}

File: integrations/trainer/tenzro_trainer/quantization.py
"""Blockwise symmetric gradient quantization — byte-identical to the Rust codec.

Mirrors ``crates/tenzro-training/src/quantization.rs``:

* **None** — raw little-endian ``f32`` values, no compression.
* **Int8** — per-block: 4-byte LE ``f32`` scale (``max_abs / 127``; an
  all-zero block encodes scale ``0.0``) followed by one ``i8`` code per value
  (``round(v / scale)`` clamped to ``[-127, 127]``). 4x smaller than f32.
* **Int4** — per-block: 4-byte LE ``f32`` scale (``max_abs / 7``) followed by
  4-bit two's-complement codes in ``[-7, 7]`` packed two per byte, low nibble
  first, odd tail zero-padded. ~8x smaller than f32.

Byte parity notes (locked by ``tests/test_quantization.py``):

* The scale is an IEEE-754 single-precision division (``f32 / f32``), so we
  compute it with ``numpy.float32`` operands before serializing.
* Rust's ``f32::round`` is round-half-away-from-zero; numpy's ``np.round``
  is banker's rounding, so we round explicitly via
  ``sign(q) * floor(|q| + 0.5)`` in float64 (exact for any f32 input).
"""

from __future__ import annotations

import struct

import numpy as np

from tenzro_trainer.types import GradientQuantization


def _round_half_away_from_zero(q: np.ndarray) -> np.ndarray:
    """Rust ``f32::round`` semantics (half away from zero), exact for f32 input."""
    q64 = q.astype(np.float64)
    return np.sign(q64) * np.floor(np.abs(q64) + 0.5)


def _iter_blocks(n: int, block_size: int) -> list[tuple[int, int]]:
    block = max(block_size, 1)
    return [(start, min(start + block, n)) for start in range(0, n, block)]


def quantize(values: np.ndarray, spec: GradientQuantization) -> bytes:
    """Encode float32 values under ``spec``. Matches Rust ``quantize`` byte-for-byte."""
    vals = np.ascontiguousarray(np.asarray(values, dtype=np.float32).reshape(-1))

    if spec.kind == "None":
        return vals.astype("<f4").tobytes()

    out = bytearray()
    for start, end in _iter_blocks(vals.size, spec.block_size):
        chunk = vals[start:end]
        max_abs = np.float32(np.max(np.abs(chunk))) if chunk.size else np.float32(0.0)
        if spec.kind == "Int8":
            scale = np.float32(max_abs) / np.float32(127.0)
            out += struct.pack("<f", float(scale))
            if scale == np.float32(0.0):
                out += bytes(chunk.size)
            else:
                q = (chunk / scale).astype(np.float32)
                codes = np.clip(_round_half_away_from_zero(q), -127, 127).astype(
                    np.int8
                )
                out += codes.tobytes()
        elif spec.kind == "Int4":
            scale = np.float32(max_abs) / np.float32(7.0)
            out += struct.pack("<f", float(scale))
            if scale == np.float32(0.0):
                codes = np.zeros(chunk.size, dtype=np.int8)
            else:
                q = (chunk / scale).astype(np.float32)
                codes = np.clip(_round_half_away_from_zero(q), -7, 7).astype(np.int8)
            nibbles = codes.astype(np.uint8) & 0x0F
            if nibbles.size % 2 == 1:
                nibbles = np.concatenate([nibbles, np.zeros(1, dtype=np.uint8)])
            packed = nibbles[0::2] | (nibbles[1::2] << np.uint8(4))
            out += packed.astype(np.uint8).tobytes()
        else:
            raise ValueError(f"unrecognized quantization kind: {spec.kind!r}")
    return bytes(out)


def dequantize(
    data: bytes, spec: GradientQuantization, expected_len: int
) -> np.ndarray:
    """Decode ``data`` into ``expected_len`` float32 values.

    Strict length validation mirrors the Rust ``dequantize`` — any byte-count
    mismatch raises :class:`ValueError` before decoding.
    """
    if expected_len < 0:
        raise ValueError("expected_len must be non-negative")

    if spec.kind == "None":
        if len(data) != expected_len * 4:
            raise ValueError(
                f"raw f32 payload must be {expected_len * 4} bytes, got {len(data)}"
            )
        return np.frombuffer(data, dtype="<f4").astype(np.float32)

    blocks = _iter_blocks(expected_len, spec.block_size)

    if spec.kind == "Int8":
        expected_bytes = expected_len + len(blocks) * 4
        if len(data) != expected_bytes:
            raise ValueError(
                f"int8 payload must be {expected_bytes} bytes for "
                f"{expected_len} values, got {len(data)}"
            )
        out = np.empty(expected_len, dtype=np.float32)
        cursor = 0
        for start, end in blocks:
            n = end - start
            scale = np.float32(struct.unpack("<f", data[cursor : cursor + 4])[0])
            cursor += 4
            codes = np.frombuffer(data, dtype=np.int8, count=n, offset=cursor)
            cursor += n
            out[start:end] = codes.astype(np.float32) * scale
        return out

    if spec.kind == "Int4":
        expected_bytes = len(blocks) * 4 + sum(
            (end - start + 1) // 2 for start, end in blocks
        )
        if len(data) != expected_bytes:
            raise ValueError(
                f"int4 payload must be {expected_bytes} bytes for "
                f"{expected_len} values, got {len(data)}"
            )
        out = np.empty(expected_len, dtype=np.float32)
        cursor = 0
        for start, end in blocks:
            n = end - start
            nbytes = (n + 1) // 2
            scale = np.float32(struct.unpack("<f", data[cursor : cursor + 4])[0])
            cursor += 4
            packed = np.frombuffer(data, dtype=np.uint8, count=nbytes, offset=cursor)
            cursor += nbytes
            nibbles = np.empty(nbytes * 2, dtype=np.uint8)
            nibbles[0::2] = packed & 0x0F
            nibbles[1::2] = (packed >> np.uint8(4)) & 0x0F
            nibbles = nibbles[:n]
            # Sign-extend the 4-bit two's-complement codes.
            codes = np.where(
                nibbles & 0x08, nibbles.astype(np.int16) - 16, nibbles.astype(np.int16)
            ).astype(np.int8)
            out[start:end] = codes.astype(np.float32) * scale
        return out

    raise ValueError(f"unrecognized quantization kind: {spec.kind!r}")


def encoded_len(length: int, spec: GradientQuantization) -> int:
    """Encoded byte count for ``length`` values — mirrors Rust ``encoded_len``."""
    if length < 0:
        raise ValueError("length must be non-negative")
    if spec.kind == "None":
        return length * 4
    blocks = _iter_blocks(length, spec.block_size)
    if spec.kind == "Int8":
        return length + len(blocks) * 4
    if spec.kind == "Int4":
        return len(blocks) * 4 + sum((end - start + 1) // 2 for start, end in blocks)
    raise ValueError(f"unrecognized quantization kind: {spec.kind!r}")


__all__ = ["dequantize", "encoded_len", "quantize"]
